relay-cogs/src/cogs.rs

```diff
@@ -0,0 +1,336 @@
+            self.recorder.record(CogsMeasurement {
+                resource: self.resource,
+                feature,
+                value: Value::Time(time),
+            });
+        }
+    }
+}
+
+impl fmt::Debug for CogsToken {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("CogsToken")
+            .field("resource", &self.resource)
+            .field("features", &self.features)
+            .finish()
+    }
+}
+
+/// A collection of weighted [app features](AppFeature).
+///
+/// Used to attribute a single COGS measurement to multiple features.
+#[derive(Clone, Copy)]
+pub struct AppFeatures(EnumMap<16, AppFeature, NonZeroUsize>);
+
+impl AppFeatures {
+    /// Attributes all measurements to a single [`AppFeature`].
+    pub fn new(feature: AppFeature) -> Self {
+        Self::builder().weight(feature, 1).build()
+    }
+
+    /// Attributes the measurement to nothing.
+    pub fn none() -> Self {
+        Self::builder().build()
+    }
+
+    /// Returns an [`AppFeatures`] builder.
+    pub fn builder() -> AppFeaturesBuilder {
+        AppFeaturesBuilder(Self(Default::default()))
+    }
+
+    /// Merges two instances of [`AppFeatures`] and sums the contained weights.
+    pub fn merge(mut self, other: Self) -> Self {
+        for (feature, weight) in other.0.into_iter() {
+            if let Some(w) = self.0.get_mut(feature) {
+                *w = w.saturating_add(weight.get());
+            } else {
+                self.0.insert(feature, weight);
+            }
+        }
+
+        self
+    }
+
+    /// Returns an iterator yielding an app feature and it's associated weight
+    /// normalized to the total stored weights in the range between `0.0` and `1.0`.
+    ///
+    /// Used to divide a measurement by the stored weights.
+    pub fn weights(&self) -> impl Iterator<Item = (AppFeature, f32)> {
+        let total_weight: usize = self.0.into_iter().map(|(_, weight)| weight.get()).sum();
+
+        self.0.into_iter().filter_map(move |(feature, weight)| {
+            if total_weight == 0 {
+                return None;
+            }
+
+            let ratio = (weight.get() as f32 / total_weight as f32).clamp(0.0, 1.0);
+            Some((feature, ratio))
+        })
+    }
+}
+
+impl fmt::Debug for AppFeatures {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "AppFeatures(")?;
+
+        let mut first = true;
+        for (feature, weight) in self.weights() {
+            if !first {
+                first = false;
+                write!(f, ", ")?;
+            }
+            write!(f, "{feature:?}={weight:.2}")?;
+        }
+        write!(f, ")")
+    }
+}
+
+impl From<AppFeature> for AppFeatures {
+    fn from(value: AppFeature) -> Self {
+        Self::new(value)
+    }
+}
+
+/// A builder for [`AppFeatures`] which can be used to configure different weights per [`AppFeature`].
+pub struct AppFeaturesBuilder(AppFeatures);
```

Comment:
Going off my previous comment, I still believe we do not need a builder pattern / a wide interface here. IMO a method `CogsToken::add_weight(&mut self, feature: AppFeature, weight: f32)` could be enough.

Reply:
Ideally I don't want the token to be mutated at all, unfortunately it's necessary, that's why there is the `update`. 

I see a problem when you can configure the individual weights on the token itself that it is very easy to mess up, tokens and individual weights can be dynamically added in different places, you can forget that there was already a weight and you just add where you should've replaced or cleared some other weight.

The current interface makes this impossible, that's also why `AppFeatures` is immutable, you can only build new weights from scratch (or merge). Hence the builder to have a separate phase of mutable weights, but once created it's immutable.

Reply:
Your example becomes:

```
let mut accountant = CogsToken::new();

// ...
let features = AppFeatures::builder();
for bucket in buckets {
    let feature = AppFeature::from(bucket.namespace);
    features.attribute(feature, bucket.size());
}
accountant.update(features);

// in drop handler, accountant divides up elapsed time proportional to attributed numbers.
```